libra-node/src/main_node.rs

```diff
@@ -184,14 +134,21 @@ pub fn setup_network(
                 .add_gossip_discovery();
         }
         DiscoveryMethod::Onchain => {
-            setup_onchain_discovery(
-                &mut network_builder,
+            let (network_tx, discovery_events) =
+                onchain_discovery::network_interface::add_to_network(&mut network_builder);
+            let onchain_discovery_builder = OnchainDiscoveryBuilder::build(
```

Comment:
would it make sense to just pass network builder in here? seems like we're just moving around data types...

Reply:
Yes, we are 'just' moving around data types.  

The reason that setup_network is a part of main_node right now is explictily because we are passing network_builder into every component as a parameter.   This creates a dependency loop where setup_network -> XXX_component -> Network_builder.  Which blocks setup_network from being part of network_builder.

This is one of the points to look into #4300 for context;  specifically look for

1)  the updated versions of "add_protocol_handler" and the corresponding "get_endpoing_configuration" calls that show up in every component.  
2). the move of network_builder to a separate crate inside networking, invoked *only* by node_builder